File: app/src/main/java/io/github/pshegger/infobook/utils/IconLoadHelper.java
package io.github.pshegger.infobook.utils;

import android.content.Context;
import android.widget.ImageView;

import com.squareup.picasso.Picasso;

import java.util.HashMap;
import java.util.Map;

import io.github.pshegger.infobook.GWApiService;
import io.github.pshegger.infobook.InfoBookApplication;
import io.github.pshegger.infobook.model.FileData;
import retrofit.Callback;
import retrofit.Response;
import retrofit.Retrofit;

/**
 * Created by pshegger on 2015.10.15..
 */
public class IconLoadHelper {
    public enum IconType {
        Item, File
    }

    private static Map<IconType, Map<String, String>> urlCache = new HashMap<>();
    static {
        urlCache.put(IconType.File, new HashMap<String, String>());
        urlCache.put(IconType.Item, new HashMap<String, String>());
    }

    public static void loadIcon(IconType type, String id, ImageView target) {
        InfoBookApplication app = InfoBookApplication.getInstance();

        if (urlCache.get(type).containsKey(id)) {
            String url = urlCache.get(type).get(id);
            Picasso.with(app)
                    .load(url)
                    .into(target);

            return;
        }

        switch (type) {
            case Item:
                loadItem(id, target);
                break;
            case File:
                loadFile(app, id, target);
                break;
        }
    }

    private static void loadItem(String id, ImageView target) {

    }

    private static void loadFile(final InfoBookApplication app, final String id, final ImageView target) {
        final GWApiService service = app.getService();

        service.getItem(id).enqueue(new Callback<FileData>() {
            @Override
            public void onResponse(Response<FileData> response, Retrofit retrofit) {
                String url = response.body().getUrl();

                urlCache.get(IconType.File).put(id, url);

                Picasso.with(app)
                        .load(url)
                        .into(target);
            }

            @Override
            public void onFailure(Throwable t) {

            }
        });
    }
}

File: app/src/main/kotlin/io/github/pshegger/infobook/utils/IconLoadHelper.kt
package io.github.pshegger.infobook.utils

import android.widget.ImageView
import com.squareup.picasso.Picasso
import io.github.pshegger.infobook.InfoBookApplication
import io.github.pshegger.infobook.model.FileData
import retrofit.Callback
import retrofit.Response
import retrofit.Retrofit

/**
 * Created by pshegger on 2015. 11. 09..
 */
object IconLoadHelper {
    enum class IconType {
        Item, File
    }

    private val urlCache: Map<IconType, MutableMap<String, String>> = mapOf(
            IconType.File to hashMapOf(),
            IconType.Item to hashMapOf()
    )

    fun loadIcon(type: IconType, id: String, target: ImageView) {
        val app = InfoBookApplication.Companion.instance ?: return;

        urlCache[type]?.let { cache ->
            if (id in cache.keys) {
                val url = cache[id]
                Picasso.with(app)
                        .load(url)
                        .into(target)

                return
            }
        }

        when (type) {
            IconType.Item -> loadItem(app, target)
            IconType.File -> loadFile(app, id, target)
        }
    }

    private fun loadItem(app: InfoBookApplication, target: ImageView) {
    }

    private fun loadFile(app: InfoBookApplication, id: String, target: ImageView) {
        val service = app.service;

        service.getItem(id).enqueue(object : Callback<FileData> {
            override fun onFailure(t: Throwable?) {

            }

            override fun onResponse(response: Response<FileData>?, retrofit: Retrofit) {
                response?.body()?.let { body ->
                    urlCache[IconType.File]?.put(id, body.url)

                    Picasso.with(app)
                            .load(body.url)
                            .into(target)
                }
            }
        })
    }
}